src/shogun/mathematics/RandomNamespace.h

```diff
@@ -0,0 +1,111 @@
+#ifndef __RANDOM_NAMESPACE_H__
+#define __RANDOM_NAMESPACE_H__
+
+#include <shogun/base/SGObject.h>
+#include <shogun/lib/common.h>
+#include <shogun/lib/config.h>
+#include <shogun/mathematics/UniformIntDistribution.h>
+#include <shogun/mathematics/UniformRealDistribution.h>
+
+#include <algorithm>
+#include <iterator>
+#include <random>
+#include <utility>
+
+namespace shogun
+{
+	namespace random
+	{
+		/** Reorders the elements in [first,  last) randomly
+		 *
+		 * @param first an iterator to the first element in the range
+		 * @param last an iterator to the last element in the range
+		 */
+		template <typename RandomIt, typename PRNG>
+		static inline void shuffle(RandomIt first, RandomIt last, PRNG&& prng)
+		{
+			using diff_t =
+			    typename std::iterator_traits<RandomIt>::difference_type;
+			UniformIntDistribution<diff_t> dist;
+			diff_t n = last - first;
+			for (diff_t i = n - 1; i > 0; --i)
+				std::swap(first[i], first[dist(prng, {0, i})]);
```

Comment:
does this generate a uniform shuffling?

Simple test: Generate an array [1,2,3,4,5,6,7,8,9]. Then shuffle. Repeat this many many times and compute a histogram for each position (all digits should appear in each position equally often)

Reply:
mmmm rather check the UniformDistribution... but since that's checked, i reckon this is good to go

Reply:
too me, it is not obvious that if you use the uniform distribution and shuffle like here, the generated permutations are uniform...but maybe I am missing something

Reply:
https://en.cppreference.com/w/cpp/algorithm/random_shuffle

read the specs + (3) implementation.

Reply:
Ah cool thx